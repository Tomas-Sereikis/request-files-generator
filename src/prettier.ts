import * as prettier from 'prettier'
import { trim } from 'lodash'

const prettierrc = require('../.prettierrc.json')

export function formatCode(code: string): string {
  return trim(
    prettier.format(code, {
      parser: 'typescript',
      ...prettierrc,
    }),
    '\n',
  )
}
