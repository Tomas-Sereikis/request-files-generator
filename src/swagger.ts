import * as swaggerParser from 'swagger-parser'
import * as swaggerScheme from 'swagger-schema-official'
import { assign, flatten } from 'lodash'

export async function validate(docs: swaggerScheme.Spec): Promise<ISwaggerDocs> {
  return await swaggerParser.validate(docs)
}

export function swaggerSchemeObjectPropertiesMerge(object: ISwaggerSchemeObject): ISwaggerSchemeObjectProperties {
  const type = SwaggerSchemeType.object
  if (object.allOf) {
    const required = flatten(object.allOf.map(curr => curr.required))
    const properties = assign({} as ISwaggerSchemeObjectProperty, ...object.allOf.map(curr => curr.properties))
    return { properties, required, type }
  } else if (object.properties && object.required) {
    return { properties: object.properties, required: object.required, type }
  } else {
    return { properties: {}, required: [], type }
  }
}

export enum SwaggerSchemeType {
  object = 'object',
  boolean = 'boolean',
  array = 'array',
  string = 'string',
  number = 'number',
  integer = 'integer',
  interface = 'interface',
}

export enum SwaggerParameterIn {
  query = 'query',
  path = 'path',
  body = 'body',
}

export interface ISwaggerDocs {
  info: ISwaggerDocsInfo
  paths: ISwaggerPaths
}

export interface ISwaggerDocsInfo {
  title: string
  description?: string
}

export interface ISwaggerPaths {
  [url: string]: ISwaggerPath
}

export interface ISwaggerPath {
  [method: string]: ISwaggerPathInfo
}

export interface ISwaggerPathInfo {
  description?: string
  summary: string
  operationId: string
  parameters?: Array<ISwaggerParameterDefaults & ISwaggerSchemeObject>
  tags: string[]
  responses: ISwaggerPathResponses
}

export interface ISwaggerParameterDefaults {
  name: string
  in: SwaggerParameterIn
  schema?: ISwaggerScheme
}

export interface ISwaggerPathResponses {
  [statusCode: string]: ISwaggerPathResponse
}

export interface ISwaggerPathResponse {
  schema: ISwaggerScheme
  description: string
}

export interface ISwaggerSchemeObjectProperty {
  [key: string]: ISwaggerScheme
}

export interface ISwaggerSchemeObjectProperties {
  properties: ISwaggerSchemeObjectProperty
  required: string[]
  type: SwaggerSchemeType.object
}

export interface ISwaggerSchemeObject extends Partial<ISwaggerSchemeObjectProperties> {
  type: SwaggerSchemeType.object
  allOf?: ISwaggerSchemeObjectProperties[]
}

export interface ISwaggerSchemeBoolean {
  type: SwaggerSchemeType.boolean
}

export interface ISwaggerSchemeArray {
  type: SwaggerSchemeType.array
  items: ISwaggerScheme
}

export interface ISwaggerSchemeString {
  type: SwaggerSchemeType.string
  enum?: string[]
}

export interface ISwaggerSchemeNumber {
  type: SwaggerSchemeType.number | SwaggerSchemeType.integer
}

export type ISwaggerScheme =
  | ISwaggerSchemeObject
  | ISwaggerSchemeBoolean
  | ISwaggerSchemeArray
  | ISwaggerSchemeString
  | ISwaggerSchemeNumber
