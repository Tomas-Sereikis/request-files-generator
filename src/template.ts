import * as assert from 'assert'
import { JoiCodeGenerator } from './generators/JoiCodeGenerator'
import {
  ISwaggerPathInfo,
  ISwaggerPaths,
  ISwaggerSchemeObject,
  ISwaggerSchemeObjectProperty,
  SwaggerParameterIn,
  SwaggerSchemeType,
} from './swagger'
import { TypeScriptCodeGenerator } from './generators/TypeScriptCodeGenerator'
import { camelCase, upperFirst, template } from 'lodash'
import { formatCode } from './prettier'

export interface ICreateTemplateParams {
  paths: ISwaggerPaths
  url: string
  method: string
  statusCode: string
  template: string
}

const removeNewLines = (str: string) => str.replace(/\r?\n|\r/g, ' ')

function parametersInToObjectScheme(
  parameters: ISwaggerPathInfo['parameters'],
  inType: SwaggerParameterIn,
): { count: number; required: string[]; scheme: ISwaggerSchemeObject } {
  const properties = parameters
    .filter(curr => curr.in === inType)
    .reduce(
      (previousValue, currentValue) => {
        if (
          currentValue.in === SwaggerParameterIn.body && 
          currentValue.name === 'body' && 
          currentValue.schema && 
          currentValue.schema.type === SwaggerSchemeType.object
        ) {
          return {
            ...previousValue,
            ...currentValue.schema.properties,
          }
        } else {
          return {
            ...previousValue,
            [currentValue.name]: currentValue.schema || currentValue,
          }
        }
      },
      {} as ISwaggerSchemeObjectProperty,
    )
  const required = parameters.filter(curr => curr.in === inType && curr.required).map(curr => curr.name)
  return {
    count: Object.keys(properties).length,
    required,
    scheme: {
      type: SwaggerSchemeType.object,
      properties,
      required,
    },
  }
}

function applyUrlParamsBuilder(url: string) {
  return (variable: string) => {
    const replaced = url.replace(/{(.*?)}/g, (match, value) => `\${${variable}.path.${value}}`)
    return `\`${replaced}\``
  }
}

export function createTemplate(options: ICreateTemplateParams): string {
  const tsCodeGenerator = new TypeScriptCodeGenerator()
  const joiCodeGenerator = new JoiCodeGenerator()

  const endpoint = options.paths[options.url][options.method]
  const response = endpoint ? endpoint.responses[options.statusCode] : null

  assert(!!endpoint, 'Swagger path info is not found!')
  assert(!!response, 'Swagger response is not found!')

  const bodyObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.body)
  const queryObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.query)
  const paramsObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.path)

  let optionsScheme = {}
  let required = []
  if (bodyObjectScheme.count) {
    optionsScheme = { ...optionsScheme, body: bodyObjectScheme.scheme }
    required.push(SwaggerParameterIn.body)
  }
  if (queryObjectScheme.count) {
    optionsScheme = { ...optionsScheme, query: queryObjectScheme.scheme }
  }
  if (queryObjectScheme.count && queryObjectScheme.required.length) {
    required.push(SwaggerParameterIn.query)
  }
  if (paramsObjectScheme.count) {
    optionsScheme = { ...optionsScheme, path: paramsObjectScheme.scheme }
    required.push(SwaggerParameterIn.path)
  }

  const optionsObject = tsCodeGenerator.onInterface(optionsScheme, required)
  const templateBuilder = template(options.template)
  const operationId = camelCase(endpoint.operationId)
  const description = endpoint.description ? removeNewLines(endpoint.description) : ''

  return formatCode(
    templateBuilder({
      applyUrlParams: applyUrlParamsBuilder(options.url),
      description: description,
      containsBody: ['post', 'put', 'patch'].includes(options.method),
      containsQuery: !!queryObjectScheme.count,
      responseInterface: tsCodeGenerator.of(response!.schema),
      optionsInterface: optionsObject,
      scheme: joiCodeGenerator.of(response!.schema),
      isDelete: ['delete'].includes(options.method),
      operationId,
      method: options.method,
      upperOperationId: upperFirst(operationId),
    }),
  )
}
