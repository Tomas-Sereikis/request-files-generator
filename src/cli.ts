import * as assert from 'assert'
import * as prompts from 'prompts'
import * as path from 'path'
import * as swagger from './swagger'
import { promises as fs } from 'fs'
import { createTemplate } from './template'

const MSG_OVERWRITE_FILE = 'Would you like to overwrite {fileName}'
const MSG_ENTER_TEMPLATE_PATH = 'Enter template path'
const MSG_ENTER_OUTPUT_PATH = 'Enter output path'
const MSG_ENTER_SWAGGER_JSON_URL = 'Enter swagger.json file url'
const MSG_SELECT_PATHS = 'Select paths that you would like to use'
const MSG_SELECT_OPERATIONS = 'Select operations that you would like to use'
const MSG_SELECT_STATUS_CODE = 'Select statusCode for {operationId}'

main(process.argv.slice(2)).catch(console.error)

async function main(argv: ReadonlyArray<string>) {
  const { value: templatePath } = await prompts({
    message: MSG_ENTER_TEMPLATE_PATH,
    name: 'value',
    type: 'text',
  })

  const templateStat = await fs.stat(templatePath)
  assert(templateStat.isFile(), 'Template file dose not exist!')
  const templateContent = String(await fs.readFile(templatePath))

  const { value: outputPath } = await prompts({
    message: MSG_ENTER_OUTPUT_PATH,
    name: 'value',
    type: 'text',
  })

  const outputStat = await fs.stat(outputPath)
  assert(outputStat.isDirectory(), 'Output directory dose not exist!')

  const { value: swaggerUrl } = await prompts({
    message: MSG_ENTER_SWAGGER_JSON_URL,
    name: 'value',
    type: 'text',
  })

  const docs = await swagger.validate(swaggerUrl)
  const choicesOfPaths = Object.keys(docs.paths).map(path => ({ value: path, title: path }))
  const { value: paths } = await prompts({
    message: MSG_SELECT_PATHS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfPaths,
  })

  const choicesOfOperations = Object.keys(docs.paths)
    .filter(path => paths.includes(path))
    .map(path => Object.values(docs.paths[path]))
    .reduce(
      (previousValue, currentValue) => [...previousValue, ...currentValue.map(curr => curr.operationId)],
      [] as string[],
    )
    .map(operation => ({ value: operation, title: operation }))

  const { value: operations } = await prompts({
    message: MSG_SELECT_OPERATIONS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfOperations,
  })

  for (const url of Object.keys(docs.paths)) {
    for (const method of Object.keys(docs.paths[url])) {
      const info = docs.paths[url][method]
      if (!operations.includes(info.operationId)) {
        continue
      }
      const statusCodes = Object.keys(info.responses)
      let statusCode: string | undefined
      if (statusCodes.length === 0) {
        continue
      } else if (statusCodes.length === 1) {
        statusCode = statusCodes[0]
      } else {
        const choicesOfStatusCodes = statusCodes.map(code => ({ value: code, title: code }))
        const selection = await prompts({
          message: MSG_SELECT_STATUS_CODE.replace('{operationId}', info.operationId),
          name: 'value',
          type: 'select',
          choices: choicesOfStatusCodes,
        })
        statusCode = selection.value
      }
      const template = await createTemplate({
        template: templateContent,
        paths: docs.paths,
        url,
        method,
        statusCode: statusCode!,
      })
      const filePath = path.resolve(outputPath, `${info.operationId}.ts`)
      const filePathStat = await fs.stat(filePath).catch(() => null)
      if (filePathStat && filePathStat.isFile()) {
        const { value: overwrite } = await prompts({
          message: MSG_OVERWRITE_FILE.replace('{filePath}', filePath),
          name: 'value',
          type: 'toggle',
          initial: true,
          active: 'yes',
          inactive: 'no',
        })
        if (overwrite === 'no') {
          continue
        }
      }
      await fs.writeFile(filePath, template)
    }
  }
}
