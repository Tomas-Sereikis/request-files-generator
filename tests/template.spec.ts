import * as swagger from '../src/swagger'
import mockOfCategories from './mocks/categories'
import mockOfChannels from './mocks/channel'
import mockOfPetstore from './mocks/petstore'
import { createTemplate } from '../src/template'
import { promises as fs } from 'fs'

let templateContent: string | undefined

beforeEach(async () => {
  templateContent = String(await fs.readFile('tests/mocks/request.template'))
})

test('should create categories template', async () => {
  const docs = await swagger.validate(mockOfCategories)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/channel-categories',
    method: 'get',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})

test('should create petstore pets template', async () => {
  const docs = await swagger.validate(mockOfPetstore)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/pets',
    method: 'get',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})

test('should create petstore pets(post) template', async () => {
  const docs = await swagger.validate(mockOfPetstore)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/pets',
    method: 'post',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})

test('should create petstore pet template', async () => {
  const docs = await swagger.validate(mockOfPetstore)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/pets/{id}',
    method: 'get',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})

test('should create channel template', async () => {
  const docs = await swagger.validate(mockOfChannels)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/channels',
    method: 'get',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})

test('should create channel(post) template', async () => {
  const docs = await swagger.validate(mockOfChannels)
  const template = await createTemplate({
    template: templateContent!,
    paths: docs.paths,
    url: '/channels',
    method: 'post',
    statusCode: '200',
  })
  expect(template).toMatchSnapshot()
})
