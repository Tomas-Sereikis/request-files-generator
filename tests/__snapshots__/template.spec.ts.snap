// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create categories template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IGetChannelCategoriesOptions = {
  query?: {
    region?: string
  }
}

export type IGetChannelCategoriesData = {
  data: Array<{
    color?: string
    id: string
    region: string
    sortIndex: number
    svg: string
    title: string
  }>
  next?: boolean
}

export const GetChannelCategoriesData = joi.object().keys({
  data: joi
    .array()
    .items(
      joi.object().keys({
        color: joi.string().allow(''),
        id: joi
          .string()
          .allow('')
          .required(),
        region: joi
          .string()
          .allow('')
          .required(),
        sortIndex: joi.number().required(),
        svg: joi
          .string()
          .allow('')
          .required(),
        title: joi
          .string()
          .allow('')
          .required(),
      }),
    )
    .required(),
  next: joi.boolean(),
})

export async function getChannelCategories(options: IGetChannelCategoriesOptions): Promise<IGetChannelCategoriesData> {
  const url = \`/channel-categories\`
  const { data } = await axios.get<IGetChannelCategoriesData>(url, {
    params: options.query,
  })
  return await joi.validate(data, GetChannelCategoriesData)
}"
`;

exports[`should create channel template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IGetChannelsOptions = {
  query?: {
    categoryId?: string
    creatorId?: string
    id?: Array<string>
    private?: boolean
    region?: string
  }
}

export type IGetChannelsData = {
  data: Array<{
    avatarImageId: string
    coverImageId: string
    creatorId: string
    datetime: string
    description: string
    followers: number
    id: string
    playlistMode: 'new' | 'shuffle'
    private: boolean
    showInNews: boolean
    title: string
    website?: string
  }>
  next?: boolean
}

export const GetChannelsData = joi.object().keys({
  data: joi
    .array()
    .items(
      joi.object().keys({
        avatarImageId: joi
          .string()
          .allow('')
          .required(),
        coverImageId: joi
          .string()
          .allow('')
          .required(),
        creatorId: joi
          .string()
          .allow('')
          .required(),
        datetime: joi
          .string()
          .allow('')
          .required(),
        description: joi
          .string()
          .allow('')
          .required(),
        followers: joi.number().required(),
        id: joi
          .string()
          .allow('')
          .required(),
        playlistMode: joi
          .string()
          .valid(['new', 'shuffle'])
          .required(),
        private: joi.boolean().required(),
        showInNews: joi.boolean().required(),
        title: joi
          .string()
          .allow('')
          .required(),
        website: joi.string().allow(''),
      }),
    )
    .required(),
  next: joi.boolean(),
})

export async function getChannels(options: IGetChannelsOptions): Promise<IGetChannelsData> {
  const url = \`/channels\`
  const { data } = await axios.get<IGetChannelsData>(url, {
    params: options.query,
  })
  return await joi.validate(data, GetChannelsData)
}"
`;

exports[`should create channel(post) template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IPostChannelsOptions = {
  body: {
    avatarImageId?: string
    coverImageId?: string
    creatorId?: string
    description?: string
    playlistMode?: 'new' | 'shuffle'
    private?: boolean
    title?: string
    website?: string
  }
}

export type IPostChannelsData = {
  data: {
    avatarImageId: string
    coverImageId: string
    creatorId: string
    datetime: string
    description: string
    followers: number
    id: string
    playlistMode: 'new' | 'shuffle'
    private: boolean
    showInNews: boolean
    title: string
    website?: string
  }
  next?: boolean
}

export const PostChannelsData = joi.object().keys({
  data: joi
    .object()
    .keys({
      avatarImageId: joi
        .string()
        .allow('')
        .required(),
      coverImageId: joi
        .string()
        .allow('')
        .required(),
      creatorId: joi
        .string()
        .allow('')
        .required(),
      datetime: joi
        .string()
        .allow('')
        .required(),
      description: joi
        .string()
        .allow('')
        .required(),
      followers: joi.number().required(),
      id: joi
        .string()
        .allow('')
        .required(),
      playlistMode: joi
        .string()
        .valid(['new', 'shuffle'])
        .required(),
      private: joi.boolean().required(),
      showInNews: joi.boolean().required(),
      title: joi
        .string()
        .allow('')
        .required(),
      website: joi.string().allow(''),
    })
    .required(),
  next: joi.boolean(),
})

export async function postChannels(options: IPostChannelsOptions): Promise<IPostChannelsData> {
  const url = \`/channels\`
  const { data } = await axios.post<IPostChannelsData>(url, options.body, {})
  return await joi.validate(data, PostChannelsData)
}"
`;

exports[`should create petstore pet template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IFindPetByIdOptions = {
  path: {
    id: number
  }
}

export type IFindPetByIdData = {
  name: string
  tag?: string
  id: number
}

export const FindPetByIdData = joi.object().keys({
  name: joi
    .string()
    .allow('')
    .required(),
  tag: joi.string().allow(''),
  id: joi.number().required(),
})

// Returns a user based on a single ID, if the user does not have access to the pet
export async function findPetById(options: IFindPetByIdOptions): Promise<IFindPetByIdData> {
  const url = \`/pets/\${options.path.id}\`
  const { data } = await axios.get<IFindPetByIdData>(url, {})
  return await joi.validate(data, FindPetByIdData)
}"
`;

exports[`should create petstore pets template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IFindPetsOptions = {
  query?: {
    tags?: Array<string>
    limit?: number
  }
}

export type IFindPetsData = Array<{
  name: string
  tag?: string
  id: number
}>

export const FindPetsData = joi.array().items(
  joi.object().keys({
    name: joi
      .string()
      .allow('')
      .required(),
    tag: joi.string().allow(''),
    id: joi.number().required(),
  }),
)

// Returns all pets from the system that the user has access to Nam sed condimentum est.
export async function findPets(options: IFindPetsOptions): Promise<IFindPetsData> {
  const url = \`/pets\`
  const { data } = await axios.get<IFindPetsData>(url, {
    params: options.query,
  })
  return await joi.validate(data, FindPetsData)
}"
`;

exports[`should create petstore pets(post) template 1`] = `
"import * as joi from 'joi'
import axios from 'axios'

export type IAddPetOptions = {
  body: {
    pet: {
      name: string
      tag?: string
    }
  }
}

export type IAddPetData = {
  name: string
  tag?: string
  id: number
}

export const AddPetData = joi.object().keys({
  name: joi
    .string()
    .allow('')
    .required(),
  tag: joi.string().allow(''),
  id: joi.number().required(),
})

// Creates a new pet in the store.  Duplicates are allowed
export async function addPet(options: IAddPetOptions): Promise<IAddPetData> {
  const url = \`/pets\`
  const { data } = await axios.post<IAddPetData>(url, options.body, {})
  return await joi.validate(data, AddPetData)
}"
`;
